consensus/hotstuff/pacemaker/pacemaker.go

```diff
@@ -12,111 +12,152 @@ import (
 // timeout.
 // To get the timeout for the next timeout, you need to call TimeoutChannel() again.
-func (p *NitroPaceMaker) TimeoutChannel() <-chan time.Time {
+func (p *ActivePaceMaker) TimeoutChannel() <-chan time.Time {
 	return p.timeoutControl.Channel()
 }
 
 // ProcessQC notifies the pacemaker with a new QC, which might allow pacemaker to
-// fast forward its view.
-func (p *NitroPaceMaker) ProcessQC(qc *flow.QuorumCertificate) (*model.NewViewEvent, bool) {
-	if qc.View < p.currentView {
-		return nil, false
+// fast-forward its view.
+func (p *ActivePaceMaker) ProcessQC(qc *flow.QuorumCertificate) (*model.NewViewEvent, error) {
+	if qc.View < p.CurView() {
+		return nil, nil
 	}
+
+	p.timeoutControl.OnProgressBeforeTimeout()
+
 	// qc.view = p.currentView + k for k ≥ 0
 	// 2/3 of replicas have already voted for round p.currentView + k, hence proceeded past currentView
 	// => 2/3 of replicas are at least in view qc.view + 1.
 	// => replica can skip ahead to view qc.view + 1
-	p.timeoutControl.OnProgressBeforeTimeout()
-
 	newView := qc.View + 1
+	err := p.updateLivenessData(newView, qc, nil)
+	if err != nil {
+		return nil, err
+	}
+
 	p.notifier.OnQcTriggeredViewChange(qc, newView)
-	return p.gotoView(newView), true
+	return &model.NewViewEvent{View: newView}, nil
 }
 
-func (p *NitroPaceMaker) ProcessTC(tc *flow.TimeoutCertificate) (*model.NewViewEvent, bool) {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) ProcessTC(tc *flow.TimeoutCertificate) (*model.NewViewEvent, error) {
+	if tc == nil || tc.View < p.CurView() {
+		return nil, nil
+	}
+
+	newView := tc.View + 1
+	err := p.updateLivenessData(newView, tc.NewestQC, tc)
+	if err != nil {
+		return nil, err
+	}
+
+	p.notifier.OnTcTriggeredViewChange(tc, newView)
+	return &model.NewViewEvent{View: newView}, nil
 }
 
-func (p *NitroPaceMaker) OnPartialTC(curView uint64) {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) OnPartialTC(newView uint64) {
+	if p.CurView() == newView {
+		p.timeoutControl.TriggerTimeout()
+	}
 }
 
 // NewestQC returns QC with the highest view discovered by PaceMaker.
-func (p *NitroPaceMaker) NewestQC() *flow.QuorumCertificate {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) NewestQC() *flow.QuorumCertificate {
+	return p.livenessData.NewestQC
 }
 
 // LastViewTC returns TC for last view, this could be nil if previous round
 // has entered with a QC.
-func (p *NitroPaceMaker) LastViewTC() *flow.TimeoutCertificate {
-	panic("not yet implemented")
+func (p *ActivePaceMaker) LastViewTC() *flow.TimeoutCertificate {
+	return p.livenessData.LastViewTC
 }
 
 // Start starts the pacemaker
-func (p *NitroPaceMaker) Start() {
+func (p *ActivePaceMaker) Start() {
 	if p.started.Swap(true) {
 		return
 	}
-	timerInfo := p.timeoutControl.StartTimeout(model.ReplicaTimeout, p.currentView)
+	timerInfo := p.timeoutControl.StartTimeout(model.ReplicaTimeout, p.CurView())
```

Comment:
We always use `ReplicaTimeout` here. What if we happen to be the leader for the first view after we start up? 

I guess the impact if any is small, because restarts are infrequent and at most this would cause one additional timed out one view per restart. But would using the more liberal timeout following a restart make sense?

Reply:
In my opinion, the two different timeouts (`ReplicaTimeout` vs `VoteCollectionTimeout`) is legacy from the passive pacemaker. We should probably remove it. 
With the active pacemaker, there are now strict requirements for a node to change view: they need proof as either a QC or TC. There is no difference anymore between the leader vs other replicas. They all follow the same rules. Except that the leader constructs the needed evidence for view change first (triggering itself to enter the view where it is primary); and then the primary broadcasts the proof (specifically the QC) as part of its block, which allows all other nodes to change the view.